flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/AbstractFineGrainedSlotManagerITCase.java

```diff
@@ -0,0 +1,672 @@
+                                            .get();
+                            assertEquals(
+                                    "The slot has not been allocated to the expected job id.",
+                                    resourceRequirements2.getJobId(),
+                                    slot.getJobId());
+                        });
+            }
+        };
+    }
+
+    /**
+     * Tests that we only request new resources/containers once we have assigned all pending task
+     * managers.
+     */
+    @Test
+    public void testRequestNewResources() throws Exception {
+        final AtomicInteger resourceRequests = new AtomicInteger(0);
+        final JobID jobId = new JobID();
+
+        new Context() {
+            {
+                resourceActionsBuilder.setAllocateResourceFunction(
+                        ignored -> {
+                            resourceRequests.incrementAndGet();
+                            return true;
+                        });
+                runTest(
+                        () -> {
+                            // the first requirements should be fulfillable with the pending task
+                            // managers of the first allocation
+                            getSlotManager()
+                                    .processResourceRequirements(
+                                            createResourceRequirements(
+                                                    jobId, getDefaultNumberSlotsPerWorker()));
+                            assertThat(resourceRequests.get(), is(1));
+
+                            getSlotManager()
+                                    .processResourceRequirements(
+                                            createResourceRequirements(
+                                                    jobId, getDefaultNumberSlotsPerWorker() + 1));
+                            assertThat(resourceRequests.get(), is(2));
+                        });
+            }
+        };
+    }
+
+    /**
+     * Test that the slot manager only allocates new workers if their worker spec can fulfill the
+     * requested resource profile.
+     */
+    @Test
+    public void testWorkerOnlyAllocatedIfRequestedSlotCouldBeFulfilled() throws Exception {
+        final AtomicInteger resourceRequests = new AtomicInteger(0);
+
+        new Context() {
+            {
+                resourceActionsBuilder.setAllocateResourceFunction(
+                        ignored -> {
+                            resourceRequests.incrementAndGet();
+                            return true;
+                        });
+                runTest(
+                        () -> {
+                            getSlotManager()
+                                    .processResourceRequirements(
+                                            createResourceRequirements(
+                                                    new JobID(), 1, getLargeSlotResourceProfile()));
```

Comment:
Whether large slot profile can be fulfilled depends on the strategy. I think this test case should be moved to `FineGrainedSlotManagerITCase`.